src/charm.py

```diff
@@ -225,26 +258,18 @@ def _external_url(self) -> str:
         # are routable virtually exclusively inside the cluster (as they rely)
         # on the cluster's DNS service, while the ip address is _sometimes_
         # routable from the outside, e.g., when deploying on MicroK8s on Linux.
-        return f"http://{self.hostname}:{self._port}"
+        scheme = "https" if self.server_cert.cert else "http"
+        return f"{scheme}://{self.hostname}:{self._port}"
 
     @property
-    def scrape_jobs(self) -> list:
+    def scrape_jobs(self) -> List[Dict[str, Any]]:
```

Comment:
The new scrape job looks quite different:
- Iirc, there's a leader guard elsewhere surrounding this function call, so without the `*`-notation we'd only get this unit.
- If we do end up having ingress in place, we will have a path component and we will need to set `metrics_path` as before. Unless you refactor the charm to use `strip_prefix=True` in the ingress lib.
- `self.hostname` is `socket.getfqdn()`, so we're losing `self.ingress_per_unit.url` compared to before. When grafana agent pushes to loki. Perhaps add a comment that clearly states that we assume self-mon is always within the model?

Reply:
Until we have OB031 approved, let's keep self.ingress_per_unit.url

Reply:
Yes, you are right.

So, we have another issue here then, because when the relation to Traefik is established (`ingress_per_unit`), we set this scrape job and we are always scrapping the leader unit: 


```python
        return [
            {
                "metrics_path": f"{external_url.path}/metrics",
                "static_configs": [{"targets": [f"{external_url.hostname}:{external_url.port}"]}],
            }
        ]
```

For instance, without Traefik we have:

![imagen](https://github.com/canonical/loki-k8s-operator/assets/939888/b736a757-0443-4bc5-b899-9408b5048ffd)

![imagen](https://github.com/canonical/loki-k8s-operator/assets/939888/e5bdb27f-c7fd-4911-b12f-45edb5496de7)


With Traefik:

![imagen](https://github.com/canonical/loki-k8s-operator/assets/939888/525abcb9-06da-4590-ab27-0609f22043d5)

![imagen](https://github.com/canonical/loki-k8s-operator/assets/939888/ccd25111-046f-4b4a-9f21-69761381fbc2)

Reply:
I imagine the peer data trick should do it.